### ISSUE TITLE
#2872: #2872: RFC: Public API

### ISSUE THREAD

**Issue by KiK0S** [datagrok-ai/public#2872]
The goal of the document is to provide context to public api initiative.

### Context and goals:

Datagrok currently has a server api and js api, which are not intended for external use. However, there are client needs for integration with datagrok. For example, ability to upload dataframes and set up dashboards programmatically.

We aim to create a public REST API and a wrapper in python so clients are able to include datagrok in their data pipelines.

#### Use case 1: pipeline-generated dashboards
Tim has a pandas dataframe that he wants to upload to Datagrok, apply some visualization, and share with other users as part of the data pipeline. With python wrapper he will be able to upload the dataframe to Datagrok with his favorite configuration previously exported in “.layout” format.

#### Use case 2: cross-language scripts
Steve has written a Datagrok R script, and shared it with the users so it can be interactively used via Datagrok. Now, he wants to be able to easily call that script from his Python code.
#### Use case 3: parameterized queries
Ed has created a connection to the database, and wrote a number of pretty complex parameterized queries. He wants to let data scientists make the same queries from Python. 
#### Use case 4: files
Ability to access and manipulate files in the file shares

### Authentication
Authentication will use API key taken from user summary (eg http://localhost:8889/u/admin/summary)
This key will be passed in the Authorization header as a Bearer token.

### Routing
All public requests will be going through /public/v1/* endpoints so they can be easily distinguished and versioned.
##Work with files
Tables will use namespaced names for identification. Example can be “project.dashboard.table”
For files we should pass a connection identifier and path to a file. Example can be “System.DemoFiles/chem%2Fsmiles.csv”

In v1, we support single table download/upload, so it can be done with uploading a single csv/parquet file. In future we can extend this with downloading/uploading in zip 

When tables are uploaded, a dashboard can be created as a set of tables and a layout file.

## Functions
To enable invoking functions, the platform should be able to identify a function by its namespaced name. As for passing arguments, we plan to support primitive data types as well as columns/data frames through their identifiers, matching the upload/download identifiers

### OpenAPI
We will expose API with openAPI specification so it can be imported easily to clients projects with codegen.

## Planned API v1

```
HTTP GET /public/v1/tables/{name} - download table by namespaced `name`
HTTP GET /public/v1/tables/{connector}/{path} - download table from `connector` located in `path`
HTTP POST /public/v1/tables/{name} - upload table as csv/parquet body and save it to `name`, 
HTTP POST /public/v1/tables/{connector}/{path} - upload table as csv/parquet body and save it to `connector` under `path`, 
HTTP POST /public/v1/tables/share/{id}/{group}?mode="view/edit" - share table `id` with `group` and give permission to view/edit based on mode.

HTTP POST /public/v1/dashboards/{name}/{table_id1},{table_id2},{...} - create dashboard `name` that has tables `table_1`, `table_2`, .... Layout can be uploaded in body.

HTTP POST /public/v1/call/{name} - call function `name` with arguments passed in JSON key-value body.

HTTP GET /public/v1/files/{connector}/{path} - download file from `connector` located in `path`
HTTP POST /public/v1/files/{connector}/{path} - upload file to `connector` under  `path`
```

## Python wrapper example

```python
import datagrok as dg
from sklearn.datasets import load_titanic


dg.token = 'Bearer …'
dg.url = 'http://localhost:8082'


### === FILES ===


# identify everything with nqName
# if we do python wrapper, result probably should be pandas
df = dg.download_table('System:DemoFiles/chem/smiles.csv')
# # or
# df = dg.download_table('clinicalcase.clindemofiles2/tv')


# some processing
df[0, 'Severity'] = 'High'

# Upload tables to either personal or public namespace
# Can also be saved to connector
df_id = dg.upload_table(df, 'smiles-new')
titanic_id = dg.upload_table(load_titanic(), 'System:DemoFiles/titanic.csv')

# Create a dashboard from dataframes with a given layout
dashboard_id = dg.upload_dashboard([df_id, titanic_id], prefix_name='core:api_demo',
                        layout='pretty.layout')

# sharing with usergroup or users, specifying namespace if needed
dg.share(dashboard_id, 'Researchers', access='edit')
dg.share(dashboard_id, 'core:Kamelichev', access='view')


### === CALLS ===


# params are serialized as json
dg.call('UsageAnalysis:TestCount', {
 'date': '21-05-2024'
})


# Column, dataframe, column list can be passed as arguments
# We can return dataframes and columns from calls
extracted_df = dg.call('core:ExtractColumns', {
 'df': df_id,
 'columns': ['USUBJID', 'AGE']
})
calculated_column = dg.call('core:ExtractValue', {
 'df': titanic_id,
 'column': 'AGE',
 'function': 'core:Abs'
})

'''
# Some fancy wrappers could also be done
proj = dg.project([df], layout='layouts/pretty_one.layout')
proj.add(datasets.load_titanic())
proj.addViewer(DG.VIEWER.BAR_CHART, {
 "valueColumnName": "SUBJID",
 "valueAggrType": "count",
 "barSortType": "by value",
 "barSortOrder": "desc",
 "splitColumnName": "DTHFL"
})
'''



```

**Comment by dnillovna** [comment]
This issue has been mirrored in Jira: https://reddata.atlassian.net/browse/GROK-15771

**Comment by dnillovna** [comment]
This issue has been mirrored in Jira: https://reddata.atlassian.net/browse/GROK-16189

**Comment by KiK0S** [comment]
Is available in 1.19
